### DISCUSSION TITLE
Tasmota LVGL:  'lv_event' object has no attribute 'code'

### DISCUSSION THREAD

**Discussion by jkwim** [arendst/Tasmota#22608] [Q&A]
I am trying to get the Berry script from https://templates.blakadder.com/sunton_ESP32-2432S028.html working on a ESP32-2424S012 device with GC9A01 Display & CST816S Touch Sensor.

Running a custom compiled Tasmota with CST816S Driver: Tasmota 14.3.0.7 (tasmota32c3-lvgl-cst816s) 

I see the following error on the console:

```
00:00:01.730 BRY: Exception> 'attribute_error' - the 'lv_event' object has no attribute 'code'
00:00:01.732 stack traceback:
00:00:01.733 	<unknown source>: in function `dropdown_changed_cb`
00:00:01.735 	<unknown source>: in function `lvgl_event_dispatch`
00:00:01.737 	<unknown source>: in function `<lambda>`

```

My investigation points to following piece of code where the problem is:

```
def dropdown_changed_cb(obj, event)
    var option = "Unknown"

    var modes = ['Auto', 'Boost', 'On', 'Off', 'Adv', 'Day' ]

    var code = event.code <<<<<<<<<<<<<<<

    print ("Dropdown code:", code)
    if code == lv.EVENT_VALUE_CHANGED
       print("Dropdown code :", code )
       option = obj.get_selected()
   
       print("Dropdown set:", modes[option])
   
       #tasmota.publish("tele/ESP32-2432S028/SENSOR/Dropdown", modes[option] )
     end

end
```

Probably the `var code = event.code` statement is throwing out the error.

When I remark the line the error is not shown.

I need some help to correct the code.

![image](https://github.com/user-attachments/assets/9e1ec46f-9730-4081-b7d2-001065b304a4)

**Comment by jkwim** [comment]
For reference here is the LVGL Documentation:
https://docs.lvgl.io/8.3/overview/event.html#event-codes

**Comment by s-hadinger** [comment]
There was a breaking change in LVGL 9.0. You should now use `event.get_code()` instead of `event.code`

**Reply by jkwim** [reply]
Oh! Thanks for the hint.

It was my mistake.

Having looked at https://tasmota.github.io/docs/LVGL_API_Reference/#class-lvevent I tried `code = event.get_code` without the braces and it was throwing out more errors, obviously.

Now it works.

**Comment by jkwim** [comment]
One more question:

Is there a way to change the alignment of the following widgets?

I want to get the clock display centered on the top line.

```
#- display wifi strength indicator icon (for professionals ;) -#
wifi_icon = lv_wifi_arcs_icon(stat_line)    # the widget takes care of positioning and driver stuff
clock_icon = lv_clock_icon(stat_line)

```

**Reply by s-hadinger** [reply]
Yes, one way to do it is using flex layout. Or simply set an absolute position since you have a fixed size screen

**Reply by jkwim** [reply]
Thanks. I was confused with the comment given for the wifi_icon above.

Managed to achieve what I wanted with absolute positioning.

```
#- display clock icon -#
clock_icon = lv_clock_icon(stat_line)
clock_icon.set_x(75)
clock_icon.set_y(10)

#- display wifi strength indicator icon (for professionals ;) -#
wifi_icon = lv_wifi_arcs_icon(stat_line)    # the widget takes care of positioning and driver stuff
wifi_icon.set_x(137)
wifi_icon.set_y(6)

```

![image](https://github.com/user-attachments/assets/4637a45e-b846-48b0-8c20-f7bd32767968)

**Comment by jkwim** [comment]
Need more help in following the example that I started with:

When I remove the remark from following line there is an error:

`# label.set_text(str(value))  <<<<<<<<<<<<<<<<<<<
`   


`
19:55:06.565 BRY: failed to load '/autoexec.be' (syntax_error - /autoexec.be:132: 'label' undeclared (first use in this function))
`

```
def slider_event_cb(obj, event)
    var slider = "Unknown"
    var value = "999"

    if   obj == slider  slider= "SLIDER"
    # elif obj == next_slider  slider= "Next"
    # elif obj == home_btn  slider= "Home"
    end

    value= obj.get_value()

    # Refresh the text
    # label.set_text(str(value))  <<<<<<<<<<<<<<<<<<<
   
    print("slider set:", str(value))
   
    tasmota.publish("tele/ESP32-2432S028/SENSOR/Slider", str(value))
end

# Create a slider in the center of the display
slider = lv.slider(scr)
slider.set_pos(20,vres-80) 
slider.set_width(200)                                              # Set the width
slider.center()                                                    # Align to the center of the parent (screen)
slider.add_event_cb(slider_event_cb, lv.EVENT_VALUE_CHANGED, None) # Assign an event function

# Create a label below the slider
label = lv.label(scr)
label.set_text("0")
label.align_to(slider, lv.ALIGN_OUT_TOP_MID, 0, -15)               # Align below the slider

```

If I move the line
`label = lv.label(scr)`

before the line
`def slider_event_cb(obj, event)`

then it works.

Is there some order of precedence that the object of the callback function can be defined after the callback function but all other objects need to be defined earlier in the order?

**Reply by sfromis** [reply]
Has nothing to do with precedence or callbacks, but the more basic Berry requirement of global variables needing to be defined before you try to use them, in source code order. What happens is that inside of a function, variables are by definition local unless already existing as globals, and local variables needs to be declared first. 

One way (outside of moving lines around) would (inside of the function) be to have:
`global.label.set_text(str(value))`
This makes it explicit to the Berry compile that `label` is a global variable, and that it cannot check if it is declared already, but just allowing the code to be compiled.
It would also work (again without moving lines around) to add a `var label` before the `def` of the callback function, to have the variable declared at the global scope. 

What I'd personally *prefer* would be to not use global variables, but refactor into a class with instance variables for shared context. That would be more of a change, but better programming practice than that example code. Of course, that preference comes from working with code on a larger scale than very simple scripting-style stuff.

**Reply by jkwim** [reply]
I see.

I just moved ahead with my testing with above change.

Now I am struggling to map the correct Berry instantiations following the C Code as a starting point.

I want to use the widget lv.arc.

Looked at the LVGL Docs for instantiating lv.arc

`lv_obj_t * arc = lv_arc_create(lv_screen_active());`

Tried in Berry:
`arc = lv.arc_create(scr)
`
This is what I get:
`22:32:01.274 BRY: failed to run compiled code (attribute_error - module 'lv' has no member 'arc_create')`

I mean I could not find any examples of` lv.arc` in Berrry but Berry API reference lists` lv.arc`:

https://tasmota.github.io/docs/LVGL_API_Reference/#widget-lvarc

What am I doing wrong here?

**Reply by jkwim** [reply]
Update:

`arc=lv.arc(src)` works.

trying with rest of the code:
```

void lv_example_arc_1(void)
{
    lv_obj_t * label = lv_label_create(lv_screen_active());

    /*Create an Arc*/
    lv_obj_t * arc = lv_arc_create(lv_screen_active());
    lv_obj_set_size(arc, 150, 150);
    lv_arc_set_rotation(arc, 135);
    lv_arc_set_bg_angles(arc, 0, 270);
    lv_arc_set_value(arc, 10);
    lv_obj_center(arc);
    lv_obj_add_event_cb(arc, value_changed_event_cb, LV_EVENT_VALUE_CHANGED, label);

    /*Manually update the label for the first time*/
    lv_obj_send_event(arc, LV_EVENT_VALUE_CHANGED, NULL);
}

```

Stuck at 
`lv_obj_set_size(arc, 150, 150);
`
API Reference does not list `set_size` either

without the set_size anyway an arc get drawn but I cannot find a way to change its radius or position.

**Reply by jkwim** [reply]
![image](https://github.com/user-attachments/assets/a155372a-0872-4b35-b857-6861128a43f4)

**Comment by s-hadinger** [comment]
Have you tried HASPmota ? It should make things much simpler for you

https://tasmota.github.io/docs/HASPmota/

**Reply by jkwim** [reply]
This is a device without PSRAM. 

Wouldn't it be more memory efficient to run raw LVGL instead of HASPmota?

Anyway I want to try HASPmota as well.

This is the first time I am trying out Berry (and LVGL also). So I am trying to get the hang of it first.

The learning curve is steep but I am making progress with help from all of you.


I managed to get an arc on the screen.
```

arc = lv.arc(scr)
arc.set_size(235,235)
arc.set_angles(20,160)
arc.set_bg_angles(20,160)
arc.center()
```

However once the arc is activated, the dropdown list and the slider does not work. Wonder whether it is a memory issue.

**With Dropdown, Slider and Arc:**
![image](https://github.com/user-attachments/assets/c0d555d2-6fb5-43e6-907a-9941d9ab23d9)

**Without Arc** 
![image](https://github.com/user-attachments/assets/4d7613fb-a840-4a4b-86fc-0f44e995c76e)

{UPDATE: Above memory discussion is not  quite correct as I understood later. See my next reply]

I hope to have multiple screens ie. multiple .be files.

What is the best way to release the memory used by the first page before I draw the second page?

**Reply by sfromis** [reply]
No, using HASPmota on top of LVGL would not reduce memory needs, on your unsuitable device without PSRAM, which is severely limiting how much stuff you can do.

**Reply by jkwim** [reply]
Just a question based on my observation:

If I re-execute the same `autoexec.be` over and over again using the command  `Br load('autoexec.be')` the free memory decreases which suggest that the previous lvgl instance doe not get destroyed.

So it appears that repeating the lines 
```
#- start LVGL and init environment -#
lv.start()
```
again will instantiate another lvgl instance.

Suppose I want to jump to a second Berry file eg: page3.be but have a different set of objects in that page, how do I destroy the objects created in the first page?

There is a hint in the following post about a command` lv.obj_clean() `but I do not understand how it works.

https://github.com/arendst/Tasmota/discussions/18597

My interpretation is that if I have` lv.start()` once in the` autoexec.be` file and continue on to the second page` page2.be` and just keep adding objects then the memory usage increase will be incremental right? Can I use the  `lv.obj_clean()` command in the `page2.be` to destroy the objects created in `autoexec.be`?

**Reply by sfromis** [reply]
Well, just don't run the same Berry code file again and again, unless your code takes step to avoid memory leaks. And if you want multiple LVGL pages, that should not be by expecting to remove all traces of one page, and start from cold with another. The idea is that the same setup can handle the pages in a combined manner. This is unrelated to what file various bits of the code happens to be stored in. Unless you create more and more additional objects, you should not expect memory usage to keep increasing, but get to a plateau with a stable set.

If you build Tasmota for development, you can get a big red "Restart Berry VM" button on the Berry REPL page, but that's obviously not for "normal operation", and not a "guarantee" for "totally zero".

Again, you do need a better device (with PSRAM) *when* you want more than basic things with a display.

**Comment by s-hadinger** [comment]
Keep in mind there is no connection between the Berry Garbage Collector and LVGL. It means that if you call repeatadly `lv.start()`, the previous LVGL objects are still using memory.

The best way to have multiple screens, just create separate screens and switch them at runtime

**Reply by jkwim** [reply]
> The best way to have multiple screens, just create separate screens and switch them at runtime

Yes, that is what I am thinking of now having read more documentation which is somewhat scattered. I mean I ended up reading LVGL Docs and Berry Docs trying to match which is where.

I now have some better understanding about how the Berry widgets are built. So will spend some time playing with them to get the hang of it. Did not want to jump to HASPmota straight away. I am likely to settle down with HASPmota with another device that I bought.

**Reply by jkwim** [reply]
Found the two methods:
```
void lv_obj_delete([lv_obj_t])
	Delete an object and all of its children. Also remove the objects from their group and remove all animations (if any). Send LV_EVENT_DELETED to deleted objects. 
		Parameters:
			obj -- pointer to an object 
void lv_obj_clean([lv_obj_t])
	Delete all children of an object. Also remove the objects from their group and remove all animations (if any). Send LV_EVENT_DELETED to deleted objects. 
		Parameters:
			obj -- pointer to an object
```

**Comment by jkwim** [comment]
I have done some basic testing to achieve the page navigation as follows:

```
autoexec.be
     |
     |
     v
  page1.be --> page2.be--> page3.be --> page4.be
     ^                                     |
     |                                     |
     |                                     |
     ---------------------------------------
```

**`autoexec.be`:**
Contains the main top status bar and the navigation buttons

`pageX.be`  contain the other widgets that I want to have.

I confirmed that the  following two commands can be used for removing the widgets as one navigates between pages.

`obj.clean` Delete all children of an object
`obj.delete` Delete an object and all of its children

However I have an issue when deleting the `dropdown  `widget.

I have the following on `page1.be`:


```
def dropdown_changed_cb(obj, event)
    var option = "Unknown"
    var modes = ['Auto', 'Boost', 'On', 'Off', 'Adv', 'Day' ]
    var code = event.get_code()
    if code == lv.EVENT_VALUE_CHANGED
       print("Dropdown code :", code )
       option = obj.get_selected()
 
       print("Dropdown set:", modes[option])
   
       #tasmota.publish("tele/ESP32-2432S028/SENSOR/Dropdown", modes[option] )
     end
end

var modes = ['Auto', 'Boost', 'On', 'Off', 'Adv', 'Day' ]
var modes_str = modes.concat('\n')
ddlist = lv.dropdown(scr)
ddlist.set_options(modes_str)
ddlist.set_pos(60,vres-200)
# ddlist.center()
ddlist.add_event_cb(dropdown_changed_cb, lv.EVENT_ALL, None)
```

And have the following in `page2.be`:

```
prev_page="page1.be"
next_page="page3.be"
label_status.set_text(next_page)
label.delete()
slider.delete()

#code.delete()
#option.delete()
print("Dropdown is being deleted")
ddlist.delete()
```

When the dropdown list get deleted I see the following error:

```
Dropdown is being deleted
BRY: Exception> 'key_error' - <ptr: 0x3fcb5204>
stack traceback:
	<native>: in native function
	<unknown source>: in function `lvgl_event_dispatch`
	<unknown source>: in function `<lambda>`
	<native>: in native function
	<unknown source>: in function `main`
	<unknown source>: in function `try_run_compiled`
	<unknown source>: in function `load`
	<unknown source>: in function `load`
	<unknown source>: in function `btn_clicked_cb`
	<unknown source>: in function `lvgl_event_dispatch`
	<unknown source>: in function `<lambda>`
BRY: failed to run compiled code (attribute_error - the 'lv_dropdown' object has no method '<')
```

From https://tasmota.github.io/docs/LVGL_API_Reference/#widget-lvobj

The base obj `lv.obj` has the delete method.


https://tasmota.github.io/docs/LVGL_API_Reference/#widget-lvdropdown

Is it a case where the the` obj.dropdown` has not inherited the methods from the base.obj?

**Reply by s-hadinger** [reply]
I will try to reproduce. It would be better to keep objects in memory (if you have enough memory) and just change the visible screen.

**Reply by jkwim** [reply]
Can you help me to understand the concept of screens in Berry LVGL please?

You are suggesting that I can load different screens in each of page1.be ... page4.be that I create right? (presumably by using a command like `screen_load(scr1)`)

I cannot find any reference to a screen widget in https://tasmota.github.io/docs/LVGL_API_Reference/#widgets either.

I saw this in LVGL documentation:

![image](https://github.com/user-attachments/assets/71853699-57cd-4ce9-8db3-4d35af1b10e9)

However in all the Berry LVGL examples I do not see the commands for creating any screens.

They all just start with selecting the active screen (whatever it may be)
`scr = lv.scr_act()            # default screen object
`
I see in LVGL documentation that you can get pointers to 4 different layers

You can get pointers to each of these screens on a specified display by using (respectively):
	• [lv_screen_active()
	• [lv_layer_top()
	• [lv_layer_sys()
         [lv_layer_bottom()

**Reply by sfromis** [reply]
Not using LVGL much, but I observe that HASPmota page creation allocates new screens like this:
`self._lv_scr = lv.obj(0)      # allocate a new screen`
And switches page (without a transition) with a function call (not a command) like this:
`lv.screen_load(self._lv_scr)`

**Reply by jkwim** [reply]
Thanks for the hints.
I have been doing lot of guess work to identify the correct syntax for creating the scr widget and your hints helped.

The correct working syntax appears to be: 
`scr1 = lv.obj(0)`
and
`lv.screen_load(scr1)`

I have decided to have the following setup for my testing with 4 page transitions.

```
autoexec.be
   (scr)
     |
     |
     v
  page1.be --> page2.be--> page3.be --> page4.be
  (scr1)         (scr2)      (scr3)      (scr4)
     ^                                     |
     |                                     |
     |                                     |
     ---------------------------------------
```

My plan is to test the following setup:

1. `scr `will have the navigation buttons (and clock and WiFi widgets) -` autoexec.be`
2. `scr1 `will have the information/widgets for `page1.be`.` page1.be` will be automatically called by `autoexec.be` so that the first page will be loaded after booting
3. `scr2` to `scr4` will contain various other information widgets

Some questions:
1. It appears that the default screen `scr` will be automatically created ie we do not have to execute `scr = lv.obj(0)`. Is my understanding correct?

2. Active Screen: What becomes the active screen? The most recently loaded screen?

3. What is the order of creating the screen and loading it?  Will this be the order?
         A.  `scr = lv.obj(0)`
         B.  `lv.screen_load(scr1)`
         C.  define the widgets that will be on the screen
or
will be B & C have to be interchanged?
         
         
4. I tried the following but failed to establish the transparency of `scr1` over `scr` so that the buttons and widgets get displayed properly. `scr1` just overwrites `scr` and the buttons do not vet visible.
    `scr` has the navigation buttons
    `scr1` has a slider widget with following screen definition:
```
    scr1 = lv.obj(0)
    lv.screen_load(scr1)
    scr1.set_style_bg_opa(lv.OPA_TRANSP, 0)
```
The `scr1` gets drawn presumably on top of `scr` and I cannot see through it . The buttons are not visible 

How do I make `scr1` to have a transparent background so that the navigation buttons are also visible?

**Reply by jkwim** [reply]
Can somebody help me please?

**Reply by s-hadinger** [reply]
Not before next week. This discussion is all over the place. Can you summarize where you are?

**Reply by jkwim** [reply]
Thanks. I can wait.

The above comment summarizes where I am.

1. I now know how to create the screen objects and attach the other child objects.
2. However I do not how exactly to control the transparency of the background of the scr object.
3. I want to define the navigation buttons and the top status bar in the default screen and then use different screen objects to show the content. I am having difficulty in understanding how to set the transparency in other screen objects so that default screen is also visible.

**Reply by s-hadinger** [reply]
Once again, I advise to use HASPmota which automatically manages this.

First, I don't understand why you have separate berry pages per screen. Does this suggests that you load new Berry pages when a button is pressed? If such, don't do that. Due to the garbage collector, it will be very hard to unload any code.

If memory permits, you should build all 4 screen objects and just display the right one according to button pressed.

**Reply by jkwim** [reply]
> First, I don't understand why you have separate berry pages per screen. Does this suggests that you load new Berry pages when a button is pressed? If such, don't do that. Due to the garbage collector, it will be very hard to unload any code.

Got your point.

> If memory permits, you should build all 4 screen objects and just display the right one according to button pressed.

OK. How can I set the transparency of a screen so that previous screen and objects on the screen also remain visible?

I want `scr` to be default and `scr1...scr4` to be drawn on top of `scr`. 
`scr` and `scr1...scr4` will not have any overlapping objects. `scrX` should be drawn with a transparent background so that objects on `scr` are also visible.

**Comment by jkwim** [comment]
I am trying to understand the MQTT handling in Berry:

Got the following function to print the JSON string from a `mqtt.subscribe`

```
import string
import mqtt
import json

def mqtt_handler(topic, idx, payload_s, payload_b)

  print("mqtt:",topic," payload:",payload_s)

  # payload: {"GridV":233.70,"BatV":54.00,"ActPw":3952.00,"PVPw":00002}
  var inveter_msg = json.load(payload_s)
  print(inveter_msg)
  print(inveter_msg['GridV']) 
 
end

```

I get the output of `print("mqtt:",topic," payload:",payload_s)` correctly:

`mqtt: stat/tasmota/SENSOR  payload: {"GridV":234.90,"BatV":54.10,"ActPw":3736.00,"PVPw":00001}`

However the following code returns errors:

```
  var inveter_msg = json.load(payload_s)
  print(inveter_msg)
  print(inveter_msg['GridV']) 
```

```
nil
BRY: Exception> 'type_error' - value 'nil' does not support subscriptable
```

It appears that the `json.load` doesn't seem to pick the `payload_s` properly.

What am I doing wrong here?


BTW Weblog 4 shows following:
```
04:21:10.671 mqtt: stat/tasmota/SENSOR  payload: {"GridV":234.10,"BatV":54.10,"ActPw":3770.00,"PVPw":00001}
04:21:10.674 nil
04:21:10.676 BRY: Exception> 'type_error' - value 'nil' does not support subscriptable
04:21:10.680 SRC: MQTT
04:21:10.681 CMD: Grp 0, Cmd 'SENSOR', Idx 1, Len 58, Pld -99, Data '{"GridV":234.10,"BatV":54.10,"ActPw":3770.00,"PVPw":00001}'
04:21:10.686 MQT: stat/tasmota/RESULT = {"Command":"Error","Input":"SENSOR {\"GridV\":234.10,\"BatV\":54.10,\"ActPw\":3770.00,\"PVP"}
04:21:16.901 BRY: GC from 24652 to 14378 bytes, objects freed 114/229 (in 1 ms) - slots from 435/548 to 243/533
```

**Reply by sfromis** [reply]
Since your `print(inveter_msg)` gives `nil`, that's because `json.load` failed due to invalid JSON, and I do se one thing which is not allowed in [JSON specs](https://www.json.org/json-en.html), leading zeroes in numbers.

One reason for that rule is that many languages use a leading zero to indicate octal data, and JSON only supports decimal. Hence the "principle of least surprise" makes it invalid to just ignore leading zeroes.

The "right" approach would be to get that bad JSON fixed, and if you're not the one constructing it, report the error to the supplier. In the real world, you could get around it with kludge code like:
```
import string
while string.find(payload_s,':0')>0 
  payload_s = string.replace(payload_s,':0',':')
end
``` 
Not nice, but not horribly bad either.

And yes, I know that this could fail if the string contains `:0` in another context, like a formatted time stamp, that's one reason to admit that it is kludge code.

**Reply by jkwim** [reply]
Wow!
Thanks for pointing it out.
It was really a co-incident that I started with with bad data to be able to catch it.
Otherwise I would have got confused with it more.

I have control over the JSON generation but those numbers are generated by my Solar Inverter. I may have to juggle a bit as I have an ESP8266 doing the serial to MQTT conversion there with Tasmota Script. At the moment the script space is quite full (I am dealing with a lot more data items from the device). That setup has worked for many years now. I have to be cautious with any modification as I have other devices which depend on those MQTT messages. Let me see.

**Reply by jkwim** [reply]
I managed to fix the JSON string at the origin. 

Actually it was my oversight as well. Other 3 values have been assigned to numeric variables (for some other processing) and I have used the numeric variables to construct the `PUBLISH` message. The last item was added later and for convenience I have picked the value directly from the serial data string which was actually a string. So I have violated the JSON spec there.

Now the values get properly extracted:

Code:
```
  var inveter_msg = json.load(payload_s)
  print(inveter_msg)
  print(inveter_msg['GridV']) 
```

Output:
```
mqtt: stat/tasmota/SENSOR  payload: {"GridV":233.10,"BatV":54.10,"ActPw":450.00,"PVPw":438.00}
{'PVPw': 438, 'GridV': 233.1, 'BatV': 54.1, 'ActPw': 450}
233.1
```

Thanks once again for spotting the issue.
